Add statuschange event

fson opened the pull request:
Add an event 'statuschange' that has two possible statuses:
* `online`: the session has been established
* `reconnecting`: the session was closed, trying to reconnect

fson committed:
Add statuschange event

Add an event 'statuschange' that has two possible statuses:
* `online`: the session has been established
* `reconnecting`: the session was closed, trying to reconnect

anp commented on index.js:
```diff
@@ -83,20 +83,30 @@ function runNgrok(opts, cb) {
+					json: true
+				});
+				cb();
+			} else if (data.obj === 'csess' && data.msg === 'session closed, starting reconnect loop') {
```
Should we maybe add tests to make sure that when we update ngrok's binary this doesn't break?

fson replied:
Now I've added tests for the `statuschange` events with `starting` and `online` statuses. I don't know how to simulate the `session closed` scenario in the test, so I punted on adding a test for that. Do you maybe have an idea how that could work?

anp replied:
I imagine we'd do something horrible like mess with ngrok's socket? I don't think we should worry about that too much though, having tests for a couple of the string parsing cases seems good enough to me!

fson replied:
Messing with the socket sounded fun, so I had to give it a quick try 😄 Getting the socket file descriptor from `lsof` was easy enough and I was able to disrupt the connection by closing the socket manually using `lldb`. I wasn't able do this reliably from a script though.

I agree having couple of test cases for output parsing is good enough for now, so I'm going to merge this.

fson committed:
Fix existing test cases

fson committed:
Disable flaky tcp tunnel tests

We only use http tunnels, so disable these tests to make the test
suite more stable.

fson committed:
Clear auth token when the test suite starts

fson committed:
Add a test case that ensures 'statuschange' events are fired
